clar2wasm/src/standard/standard.wat

```diff
@@ -255,64 +258,90 @@
+        (if (i64.eqz (i64.clz (local.get $b_hi)))
             (call $runtime-error (i32.const 0))
         )
 
         ;; res *= 2, meaning res <<= 1
-        (local.set $a_hi
+        (local.set $b_hi
             (i64.or
-                (i64.shl (local.get $a_hi) (i64.const 1))
-                (i64.shr_u (local.get $a_lo) (i64.const 63))
+                (i64.shl (local.get $b_hi) (i64.const 1))
+                (i64.shr_u (local.get $b_lo) (i64.const 63))
             )
         )
-        (local.set $a_lo (i64.shl (local.get $a_lo) (i64.const 1)))
+        (local.set $b_lo (i64.shl (local.get $b_lo) (i64.const 1)))
 
-        ;; if b is odd, we add b
-        (if (i32.wrap_i64 (i64.and (local.get $b_lo) (i64.const 1)))
+        ;; if b is odd ($tmp), we add a
+        (if (local.get $tmp)
             (then
-                (call $add-uint (local.get $a_lo) (local.get $a_hi) (local.get $b_lo) (local.get $b_hi))
-                (local.set $a_hi)
-                (local.set $a_lo)
+                (call $add-uint (local.get $b_lo) (local.get $b_hi) (local.get $a_lo) (local.get $a_hi))
+                (local.set $b_hi)
+                (local.set $b_lo)
             )
         )
-        (local.get $a_lo) (local.get $a_hi)
+        (local.get $b_lo) (local.get $b_hi)
     )
     
 
     (func $mul-int (type 1) (param $a_lo i64) (param $a_hi i64) (param $b_lo i64) (param $b_hi i64) (result i64 i64)
-        (local $expected_sign i64)
+        (local $sign i32)
 
-        ;; Shortcut if either a or b is zero
-        (if (i32.or
-                (i64.eqz (i64.or (local.get $a_lo) (local.get $a_hi)))
-                (i64.eqz (i64.or (local.get $b_lo) (local.get $b_hi))))
-            (return (i64.const 0) (i64.const 0))
+        ;; take the absolute value of the operands, and compute the expected sign in 3 steps:
+        ;; 1. Absolute value of a
+        (select
+            (i64.sub (i64.const 0) (local.get $a_lo))
+            (local.get $a_lo)
+            (local.tee $sign (i64.lt_s (local.get $a_hi) (i64.const 0))) ;; sign is the sign bit of a
         )
-
-        ;; compute the expected sign of the product
-        (local.set $expected_sign 
-            (i64.xor 
-                (i64.shr_s (local.get $a_hi) (i64.const 63)) 
-                (i64.shr_s (local.get $b_hi) (i64.const 63))
-            )
+        (select
+            (i64.sub (i64.const 0) (i64.add (local.get $a_hi) (i64.extend_i32_u (i64.ne (local.get $a_lo) (i64.const 0)))))
+            (local.get $a_hi)
+            (local.get $sign)
         )
+        ;; 2. Absolute value of b
+        (select
+            (i64.sub (i64.const 0) (local.get $b_lo))
+            (local.get $b_lo)
+            (local.tee $sign (i64.lt_s (local.get $b_hi) (i64.const 0))) ;; sign is sign bit of b now
+        )
+        (select
+            (i64.sub (i64.const 0) (i64.add (local.get $b_hi) (i64.extend_i32_u (i64.ne (local.get $b_lo) (i64.const 0)))))
```

Comment:
Can you add a comment here? I think this is carrying a 1 from the lower bits, but its a bit confusing.

Reply:
Good question! I have to admit that the absolute value function was generated from the Rust code
```rust
#[no_mangle]
fn abs(a: i128) -> i128 {
    a.abs()
}
```

I'll try to find out why that works.

Reply:
Well, I have to say, this puzzles me. But since it works, I don't intend to spend more time looking for an answer to this question.
How about a comment that says that it is generated from ... ?